### PR TITLE
Feature/user authentication

### PR DESCRIPTION
**Pull request by yong-jie**
1) Added validator and session into middleware folders.
2) Modified User schema for authentication related purposes.
3) Modified user route for API endpoint: GET /user/:username.
4) Created authentication route that currently supports: POST: /authentication/create with necessary params.

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -4,11 +4,14 @@ import Debug from 'debug';
 import logger from 'morgan';
 import cookieParser from 'cookie-parser';
 import bodyParser from 'body-parser';
+import validator from './middleware/validator';
+import session from './middleware/session';
```
Comment (marvinchin) [inline app.js]: relative imports (stuff from your own modules) should come after absolute imports (stuff like 'express')

File: app.js [hunk 2]
```diff
@@ -31,11 +34,14 @@ mongoose.connect(config.DB_CONNECTION_STRING, {}, (err) => {
 app.use(logger('dev'));
 app.use(bodyParser.json());
 app.use(bodyParser.urlencoded({ extended: false }));
+app.use(validator());
 app.use(cookieParser());
 app.use(express.static(path.join(__dirname, 'public')));
+app.use(session());
```
Comment (marvinchin) [inline app.js:40]: Are validator and session things that work hand in hand? If so may want to put validator and session together.

File: middleware/session.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import session from 'express-session';
+import connectMongo from 'connect-mongo';
+import config from '../config';
+
+const MongoStore = connectMongo(session);
+
+const initSession = () => session({
+  secret: 'test', // TODO: Store secret outside of source code.
+  store: new MongoStore({
+    host: '127.0.0.1',
```
Comment (marvinchin) [inline middleware/session.js]: this should be in config

File: middleware/validator.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+const validator = () => expressValidator({
+  customValidators: {
+    noSpaces: (str) => {
+      return !(/\s/.test(str));
+    },
+    validGender: (str) => {
+      return str === 'Male' || str === 'Female';
+    },
+    notEmpty: (str) => {
+      return str !== '' && str !== null;
```
Comment (marvinchin) [inline middleware/validator.js]: can be simplified to return str && str !== ''

File: app.js (hunk 2 above)
Comment (yong-jie) [reply app.js:40]: They work separately, so will keep them like that.

File: middleware/session.js (hunk 3 above)
Comment (yong-jie) [reply middleware/session.js]: Removed host and port, and transferred client secret into config.

**Review by marvinchin — APPROVED** [review]
Naise

**Pull request merged**